galsim/phase_psf.py

```diff
@@ -0,0 +1,1071 @@
+    j is the linear Noll coordinate, n is the radial Zernike index and m is the azimuthal Zernike
+    index.
+    @param [in] j Zernike mode Noll index
+    @return (n, m) tuple of Zernike indices
+    @see <https://oeis.org/A176988>.
+    """
+    if (j == 0):
+        raise ValueError("Noll indices start at 1, 0 is invalid.")
+
+    n = 0
+    j1 = j-1
+    while (j1 > n):
+        n += 1
+        j1 -= n
+
+    m = (-1)**j * ((n % 2) + 2 * int((j1+((n+1) % 2)) / 2.0))
+    return (n, m)
+
+
+def _zern_norm(n, m):
+    """Normalization coefficient for zernike (n, m).
+
+    Defined such that \int_{unit disc} Z(n1, m1) Z(n2, m2) dA = \pi if n1==n2 and m1==m2 else 0.0
+    """
+    if m == 0:
+        return np.sqrt(1./(n+1))
+    else:
+        return np.sqrt(1./(2.*n+2))
+
+
+def _zern_rho_coefs(n, m):
+    """Compute coefficients of radial part of Zernike (n, m).
+    """
+    kmax = (n-abs(m))/2
+    A = [0]*(n+1)
+    for k in xrange(kmax+1):
+        val = (-1)**k * _nCr(n-k, k) * _nCr(n-2*k, kmax-k) / _zern_norm(n, m)
+        A[n-2*k] = val
+    return A
+
+
+def _zern_coef_array(n, m, shape=None):
+    """Assemble coefficient array array for evaluating Zernike (n, m) as the real part of a
+    bivariate polynomial in abs(rho)^2 and rho, where rho is a complex array indicating position on
+    a unit disc.
+    """
+    if shape is None:
+        shape = ((n//2)+1, abs(m)+1)
+    out = np.zeros(shape, dtype=np.complex128)
+    coefs = np.array(_zern_rho_coefs(n, m), dtype=np.complex128)
+    if m < 0:
+        coefs *= -1j
+
+    for i, c in enumerate(coefs[abs(m)::2]):
+        out[i, abs(m)] = c
+    return out
+
+
+def horner(x, coef):
+    """Evaluate univariate polynomial using Horner's method.
+
+    I.e., take A + Bx + Cx^2 + Dx^3 and evaluate it as
+    A + x(B + x(C + x(D)))
+
+    @param x     Where to evaluate polynomial.
+    @param coef  Polynomial coefficients of increasing powers of x.
+    @returns     Polynomial evaluation.  Will take on the shape of x if x is an ndarray.
+    """
+    result = 0
+    for c in coef[::-1]:
+        result = result*x + c
+    return result
+
+
+def horner2d(x, y, coefs):
+    """Evaluate bivariate polynomial using nested Horner's method.
+
+    @param x      Where to evaluate polynomial (first covariate).  Must be same shape as y.
+    @param y      Where to evaluate polynomial (second covariate).  Must be same shape as x.
+    @param coefs  2D array-like of coefficients in increasing powers of x and y.
+                  The first axis corresponds to increasing the power of y, and the second to
+                  increasing the power of x.
+    @returns      Polynomial evaluation.  Will take on the shape of x and y if these are ndarrays.
+    """
+    result = 0
+    for coef in coefs[::-1]:
+        result = result*x + horner(y, coef)
+    return result
+
+
+class OpticalScreen(object):
+    """
+    @param aberrations   Zernike polynomial aberrations sequence in waves.
```

Comment:
> I'd recommend waiting a few days for me to whip these into better shape.

Okay, will do!  I will wait for you to post on here saying that it's ready.

> I was thinking that we could keep the existing front end to OpticalPSF, but use the new stuff for the 
> implementation. 

Are all aspects of the implementation in optics.py in here though?  I think some details are not.  So that could require some work to merge together.